cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+package machines
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// Machine represents a single Node that should exist (whether it does or
+// not yet). In this model, there is no grouping of nodes to scale with a
+// numeric field. Each Machine exists independently, and grouping can only
+// be inferred via label selectors.
+//
+// In order for a new Node to be created, one can generically create a new
+// Machine object, possibly copying the spec from an existing Machine
+// or a template. To scale down the cluster, delete specific instances of
+// Machines and the underlying Nodes will be unregistered/deprovisioned.
+// Separate provider-specific controllers will watch Machine objects they can
+// act on (like a GCE cloud controller watching for only Machines destined for
+// GCE) and take the appropriate actions.
+//
+// Any updates to the MachineSpec will be actuated to change the Node in
+// place or replace the Node with one conforming to the spec. In this model,
+// the fact that a controller is able to upgrade a node via in-place upgrades
+// or via a cloud replacement is an implementation detail without API controls.
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta
+
+	// Provider-specific serialized configuration to use during node
+	// creation. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field, akin to component config.
+	// +optional
+	ProviderConfig string
+
+	// A list consisting of "Master" and/or "Node".
+	//
+	//                 +-----------------------+------------------------+
+	//                 | Master present        | Master absent          |
+	// +---------------+-----------------------+------------------------|
+	// | Node present: | Install control plane | Join the cluster as    |
+	// |               | and be schedulable    | just a node            |
+	// |---------------+-----------------------+------------------------|
+	// | Node absent:  | Install control plane | Invalid configuration  |
+	// |               | and be unscheduleable |                        |
+	// +---------------+-----------------------+------------------------+
+	Roles []string
```

Comment:
I'd suggest just making this a string and having a single identifier for each possible configuration. It's much less confusing that way. e.g. have "SchedulableMaster," "UnschedulableMaster," and "Node."

Reply:
I actually originally had it that way in an earlier draft that was Google-internal only, and got the opposite feedback from Brian Grant, Tim Hockin, and Daniel Smith: make it a list of strings instead.

I think there might be contention over exactly what roles or node installation scenarios we want to support directly in this API, however.

Reply:
Can you ping me the internal doc so I can get up to speed on the rationale?